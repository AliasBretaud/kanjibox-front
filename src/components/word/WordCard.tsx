import type { $Kanji, $Word } from "@/types";
import { Box, Card, CardContent } from "@mui/material";
import Table, { Row } from "@/components/ui/Table";
import { isKanji } from "wanakana";
import { ClickableKanji } from "./ClickableKanji";

const getKanji = (char: string, wordKanjis?: $Kanji[]) =>
  isKanji(char) && wordKanjis?.length
    ? wordKanjis.find((k) => k.value === char)
    : null;

const WordCard = (word: $Word) => (
  <Card
    sx={{
      textAlign: "center",
      width: "100%",
      maxWidth: "300px",
    }}
  >
    <CardContent>
      <Box
        component="div"
        fontSize="35pt"
        color="black"
        sx={{ cursor: "default", opacity: 0.75 }}
      >
        <ruby>
          <Box display="table" sx={{ wordBreak: "break-word" }}>
            {word.value.split("").map((char, idx) => {
              const kanji = getKanji(char, word.kanjis);
              return kanji ? (
                <ClickableKanji key={idx} {...kanji} />
              ) : (
                <span key={idx}>{char}</span>
              );
            })}
          </Box>
          <rp>(</rp>
          <Box component="rt" fontSize="15pt">
            {word.furiganaValue}
          </Box>
          <rp>)</rp>
        </ruby>
      </Box>
      <Table>
        <Row title="翻訳" data={word.translations.en} />
      </Table>
    </CardContent>
  </Card>
);

export default WordCard;
