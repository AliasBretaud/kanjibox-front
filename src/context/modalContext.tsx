"use client";

import type { $Kanji } from "@/types";
import React from "react";

export type ModalOptions = Partial<Record<string, string | string[] | $Kanji>>;

type ContextValues =
  | {
      shownModal: string;
      setShownModal: React.Dispatch<React.SetStateAction<string>>;
      modalOptions: ModalOptions;
      setModalOptions: React.Dispatch<React.SetStateAction<ModalOptions>>;
    }
  | undefined;

export const ModalContext = React.createContext<ContextValues>(undefined);

export function ModalProvider({ children }: { children: React.ReactNode }) {
  const [shownModal, setShownModal] = React.useState<string>("");
  const [modalOptions, setModalOptions] = React.useState<ModalOptions>({});
  return (
    <ModalContext.Provider
      value={{ shownModal, setShownModal, modalOptions, setModalOptions }}
    >
      {children}
    </ModalContext.Provider>
  );
}
