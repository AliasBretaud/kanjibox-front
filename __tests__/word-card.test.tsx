import WordCard from "@/components/word/WordCard";
import { ModalProvider } from "@/context/modalContext";
import type { $Word } from "@/types/models";
import { render, screen } from "@testing-library/react";

const wordMock: $Word = {
  value: "言葉",
  furiganaValue: "ことば",
  translations: {
    en: ["Word"],
  },
};

describe("WordCard component tests", () => {
  it("renders the component", () => {
    render(
      <ModalProvider>
        <WordCard {...wordMock} locale="en" />
      </ModalProvider>,
    );

    const furiganaValue = screen.getByText("ことば");
    const translation = screen.getByText("Word");
    "言葉"
      .split("")
      .map((c) => screen.getByText(c))
      .forEach((v) => expect(v).toBeInTheDocument());
    expect(furiganaValue).toBeInTheDocument();
    expect(translation).toBeInTheDocument();
  });

  it("renders WordCard unchanged", () => {
    const { container } = render(
      <ModalProvider>
        <WordCard {...wordMock} locale="en" />
      </ModalProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
