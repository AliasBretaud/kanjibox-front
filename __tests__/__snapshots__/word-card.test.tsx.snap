// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordCard component tests renders WordCard unchanged 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 w-full max-w-md flex flex-col items-center relative css-i0bk1x-MuiPaper-root"
  >
    <div
      class="MuiBox-root css-yqcczr"
    />
    <div
      class="MuiStack-root css-12h60pv-MuiStack-root"
    >
      <span
        lang="ja"
        style="display: flex; flex-flow: row wrap; font-family: 'ヒラギノ角ゴ ProN', 'Hiragino Kaku Gothic ProN', 'TakaoPゴシック', TakaoPGothic, '游ゴシック', '游ゴシック体', YuGothic, 'Yu Gothic', 'メイリオ', Meiryo, 'ＭＳ ゴシック', 'MS Gothic', HiraKakuProN-W3, 'MotoyaLCedar', 'Droid Sans Japanese', sans-serif; font-size: 35pt; color: black; cursor: default; opacity: 0.75; justify-content: center; align-items: center; min-height: 83px; align-content: center;"
      >
        <span
          class="font-kanji"
          lang="ja"
          style="display: inline-flex; line-height: 1; flex-flow: column nowrap; justify-content: flex-end; align-items: center; align-self: flex-end; gap: 10px;"
        >
          <span
            lang="ja"
            style="display: block; font-size: 15pt; letter-spacing: -0.02em; margin: 0px 0.1em; padding-top: 0.2em; padding-bottom: 0.1em; user-select: none; opacity: 0.9;"
          >
            ことば
          </span>
          <span
            lang="ja"
            style="display: block;"
          >
            <span>
              言
            </span>
            <span>
              葉
            </span>
          </span>
        </span>
      </span>
      <table
        class="w-full border-collapse text-left font-kanji border border-solid border-neutral-300"
      >
        <tbody>
          <tr>
            <th
              class="bg-stone-100 text-center text-sm w-[75px] border border-solid border-neutral-300"
              rowspan="1"
            >
              翻訳
            </th>
            <td
              class="py-[5px] px-2 border border-solid border-neutral-300"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-vkc4v8-MuiTypography-root"
              >
                Word
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
